File: MATLAB/clipperplus.m
function [c] = clipperplus(A)

%% Inputs:
%
%       A:  an nxn unweighted, symmetric adjacency matrix. 
%% Outputs:
%
%       c:  vertices that form maximal clique

sparse_A = sparse(A);
k = core_numbers(sparse_A);

[c,cmax] = PMC(A,k);% Degeneracy-ordered greedy clique

L = find(k >= length(c)); % Find vertices with core numbers equal or more than cmax

pruned_A = A(L ,L);

 u0 = ones(1, length(L));
  common_elements = L(ismember(L,c));

 for i = 1:length(common_elements)
 u0(common_elements(i)) = 0 ;%initial guess
 end
  
 if ~isempty(pruned_A)
   u = clipper(pruned_A, u0); %Continuous-relaxation clique
   u(u > 0) = 1;
   cprime = L' .*u';
   cprime = cprime(cprime ~= 0);%vertices that form maximal clique
   
   if (length(cprime) >= length (c))
     c = cprime;
   end
end
 


end

File: cpp/clipperplus.cpp

// This C++ program takes the adjacency matrix of a graph and the core numbers of each vertex.
// It identifies the vertices that form a maximal clique in the graph and prints them using cout.

#include"clipper.hpp"
#include"pmc.hpp"
#include"set"



int main(){
  std::vector<std::vector<int>> A;
     
 A ={{0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0},
     {1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1},
     {0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0},
     {1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0},
     {0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1},
     {1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 1},
     {0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1},
     {1, 1, 0, 0, 1, 1, 1, 0, 1, 0, 1},
     {0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1},
     {1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1},
     {0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0}}; // adjacency matrix
  int n = A.size();
  std::vector<int> k; // core numbers
  k = {4,4,4,4,4,4,4,4,4,4,4};
   
    std::vector<int> C;
    C = PMC(A, k) ;

    std::vector<int> L;
   
    std::cout << std::endl;
    for (int i = 0; i < n; ++i) {
        if (k[i] >= C.size()) {
            L.push_back(i);
        }}
        
        if(!L.empty()){
      
    }
      MatrixXd pruned_A(L.size(),L.size());
      



    for (int i = 0; i < L.size(); ++i) {
        for (int j = 0; j < L.size(); ++j) {
             pruned_A(i, j) = A[L[i]][L[j]];
        }
    }
VectorXd u0 = VectorXd::Ones(L.size());

std::set<int> C_set(C.begin(), C.end());

for (int i = 0; i < L.size();++i){
    if(C_set.count(L[i])> 0){
       u0[i] = 0;
    }
 }
  
  std::vector<int> C_prime; 

   VectorXd u(pruned_A.rows());
   
   u = clipper (pruned_A , u0);
   
    for (int j = 0; j < u.size(); ++j) {
       if(u[j] > 0 ){C_prime.push_back(L[j]);}

  }
if (C_prime.size() >= C.size() ){
  C = C_prime;
}


for (auto ci : C){
  std::cout << ci << " ";
}
std::cout << std::endl;}